src/content/blog/cheerp-typescript.mdx

```diff
@@ -0,0 +1,393 @@
+nodes. It is as simple as iterating over the metadata and outputting everything
+to a file in the right format.
+
+One extra complexity with the TypeScript output is that we need to be able to
+inspect the types of function arguments and return values. The function
+signatures themselves only offer a mangled name for these types, and it's up to
+us to demangle and parse this name ourselves. For example, when we encounter
+`_ZN6client6TArrayIPNS_6StringEEE`, we must first parse this string to see that
+it's a `client::TArray` of `client::String` pointers, and then we must
+translate this into the appropriate TypeScript type, `Array<string>`.
+
+We also have some special handling for Clientlib types. The `client` namespace
+is removed, `TArray` becomes just `Array`, `_Union` becomes a TypeScript union,
+`_Function` becomes a TypeScript function, and primitive types are translated
+to their TypeScript equivalent.
+
+## From TypeScript to C++ headers
+
+Calling into C++ from JavaScript is only one side of the story. We must also be
+able to call JavaScript functions from C++. You may need to call your own
+JavaScript code, call functions of the DOM, or maybe even a third party
+library.
+
+Let's start by writing a function in JavaScript that we want to call from C++.
+
+```js title="library.js"
+function square(x) {
+    return x * x;
+}
+```
+
+With Cheerp, by also declaring this function in C++ in the `client` namespace,
+we will then be able to call it from C++.
+
+```cpp title="main.cpp"
+namespace [[cheerp::genericjs]] client {
+    double square(double x);
+}
+
+[[cheerp::genericjs]]
+int main() {
+    std::cout << client::square(2) << std::endl; // prints "4"
+}
+```
+
+```sh
+/opt/cheerp/bin/clang++ main.cpp -o main.js
+cat library.js main.js | node
+```
+
+### Automation
+
+Writing these declarations by hand has the same issues as before. It's tedious.
+We have to write down each declaration twice. And if we ever change the
+signature of `square`, we need to update the C++ signature as well.
+
+To solve this, we have developed
+[ts2cpp](https://github.com/leaningtech/ts2cpp). ts2cpp is a program that
+converts TypeScript declaration files into C++ headers that can be used with
+Cheerp.
+
+Ideally we would write the `square` function in TypeScript. But let's start
+by seeing how far we can get without touching the `square` function. Let's
+pretend it's part of some third party library that we cannot modify.
+
+We can use the TypeScript compiler to generate a declaration file from
+JavaScript code.
+
+```sh
+npx tsc library.js --allowJs --declaration --emitDeclarationOnly
+```
+```ts title="library.d.ts"
+declare function square(x: any): number;
+```
+
+Not bad! While it wasn't able to figure out the type of the function argument,
+it still got the correct return type.
+
+Now we can use ts2cpp to generate a C++ header.
+
+```sh
+ts2cpp library.d.ts -o library.h --pretty
```

Comment:
Is ts2cpp available on npm? If not, can we make it available and use npx? Or npm install it

Reply:
Added to my TODO list